File: clutch-backend/app/schemas/esports.py
"""Schémas du domaine esport — COPIE FIDÈLE de `clutch-frontend/src/types/esports.ts`.

Ne rien renommer, ne rien ajouter au JSON de sortie : le front fait foi.
"""

from typing import Literal

from app.schemas.base import ApiModel

# Valeurs EXACTES du front (esports.ts)
GameId = Literal[
    "val", "lol", "cs2", "dota", "rl", "ow",
    "apex", "r6", "pubg", "fn", "ff", "mlbb", "hok", "sf6", "tk8", "bo7", "tft",
]
MatchStatus = Literal["upcoming", "done"]
BestOf = Literal["BO1", "BO3", "BO5"]


class GameOut(ApiModel):
    """Miroir de `Game` (front)."""

    id: GameId
    name: str
    short: str
    tag: str
    bg_url: str  # JSON : bgUrl
    logo_url: str | None = None  # JSON : logoUrl
    full_logo_url: str | None = None  # JSON : fullLogoUrl


class TeamOut(ApiModel):
    """Miroir de `Team` (front)."""

    id: str
    name: str
    tag: str
    country_code: str  # JSON : countryCode
    logo_url: str | None = None  # JSON : logoUrl


class PlayerOut(ApiModel):
    """Miroir de `Player` (front) — joueur d'un roster d'équipe."""

    id: str
    name: str
    country_code: str  # JSON : countryCode
    role: str | None = None
    game_id: GameId | None = None  # JSON : gameId (effectif par jeu)


class StreamOut(ApiModel):
    """Miroir de `Stream` (front) — lien de diffusion d'un match."""

    platform: str
    url: str


class VetoStepOut(ApiModel):
    """Miroir de `VetoStep` (front) — une étape du veto des cartes."""

    order: int
    type: Literal["ban", "pick", "decider"]
    team: Literal["a", "b"] | None = None  # None pour la decider
    map: str


class MapPlayerOut(ApiModel):
    """Miroir de `MapPlayer` (front) — stats d'un joueur sur une carte."""

    side: Literal["a", "b"]
    name: str
    country_code: str  # JSON : countryCode
    kills: int
    deaths: int
    assists: int
    acs: float | None = None  # Valorant
    adr: float | None = None  # Valorant / CS2
    hs: float | None = None  # % headshots (Valorant)
    agent: str | None = None  # Valorant
    champion: str | None = None  # League of Legends
    role: str | None = None  # poste LoL ("Top", "Mid", "Support"…)
    hero: str | None = None  # Dota 2
    gpm: float | None = None  # gold/min (Dota 2)
    xpm: float | None = None  # xp/min (Dota 2)
    lasthits: float | None = None  # last hits / CS (Dota 2)  # JSON : lasthits
    networth: float | None = None  # or total (Dota 2)
    level: float | None = None  # niveau du héros (Dota 2)


class HalfOut(ApiModel):
    """Score d'une demi-temps (CS2 : ct/t ; R6 : atk/def)."""

    side: str   # "ct" | "t" | "atk" | "def"
    score: int


class OpBanOut(ApiModel):
    """Ban d'opérateur R6 : nom + phase (atk/def)."""

    name: str
    type: str   # "atk" | "def"


class MapScoreOut(ApiModel):
    """Miroir de `MapScore` (front)."""

    name: str
    score_a: int  # JSON : scoreA
    score_b: int  # JSON : scoreB
    winner: Literal["a", "b"] | None = None
    live: bool | None = None
    players: list[MapPlayerOut] | None = None  # scoreboard par joueur
    # Draft Dota 2 : picks, bans, side, durée
    heroes_a: list[str] | None = None  # JSON : heroesA
    heroes_b: list[str] | None = None  # JSON : heroesB
    bans_a: list[str] | None = None    # JSON : bansA
    bans_b: list[str] | None = None    # JSON : bansB
    side_a: str | None = None          # JSON : sideA ("radiant"/"dire")
    side_b: str | None = None          # JSON : sideB
    length: str | None = None          # durée de la partie ("31m48s")
    # CS2/R6 : mi-temps avec côté
    halves_a: list[HalfOut] | None = None   # JSON : halvesA
    halves_b: list[HalfOut] | None = None   # JSON : halvesB
    vod: str | None = None                  # lien VOD par carte
    # R6 : bans d'opérateurs
    op_bans_a: list[OpBanOut] | None = None  # JSON : opBansA
    op_bans_b: list[OpBanOut] | None = None  # JSON : opBansB


class MatchOut(ApiModel):
    """Miroir de `Match` (front).

    `date`/`time` sont calculées par le service depuis le timestamp UTC en
    base, converties dans DISPLAY_TZ. `oddsA` n'existe pas côté backend
    (décision actée : aucune valeur fabriquée).
    """

    id: str
    game_id: GameId  # JSON : gameId
    team_a: TeamOut  # JSON : teamA (équipe EMBARQUÉE, jamais un id)
    team_b: TeamOut  # JSON : teamB
    status: MatchStatus
    phase: str
    best_of: BestOf  # JSON : bestOf
    date: str  # "YYYY-MM-DD" dans DISPLAY_TZ
    time: str  # "HH:mm" dans DISPLAY_TZ
    score_a: int | None = None  # JSON : scoreA
    score_b: int | None = None  # JSON : scoreB
    result_a: str | None = None  # JSON : resultA ("W" | "L" | "FF" | "DQ")
    result_b: str | None = None  # JSON : resultB
    maps: list[MapScoreOut] | None = None
    current_map_label: str | None = None  # JSON : currentMapLabel
    viewers: str | None = None
    streams: list[StreamOut] | None = None  # liens de diffusion (streamurls LPDB)
    veto: list[VetoStepOut] | None = None  # veto des cartes (extradata.mapveto)
    # Drapeau best-effort quand LPDB suggère une victoire administrative.
    likely_forfeit: bool | None = None  # JSON : likelyForfeit
    # Roster des joueurs alignés pour ce match (chargé uniquement sur /matches/{id})
    team_a_players: list[PlayerOut] | None = None  # JSON : teamAPlayers
    team_b_players: list[PlayerOut] | None = None  # JSON : teamBPlayers

File: clutch-frontend/src/types/esports.ts
/**
 * Types du domaine esport — TEMPORAIRES.
 * Le backend n'existe pas encore : ces types miment la future réponse de
 * notre API interne. Quand la spec OpenAPI sera disponible, ils seront
 * remplacés par les types générés dans `src/api/generated/` (source de
 * vérité = le backend) et ce fichier sera supprimé.
 */

export type GameId =
  | 'val' | 'lol' | 'cs2' | 'dota' | 'rl' | 'ow'
  | 'apex' | 'r6' | 'pubg' | 'fn' | 'ff' | 'mlbb' | 'hok' | 'sf6' | 'tk8' | 'bo7' | 'tft';

export type MatchStatus = 'upcoming' | 'live' | 'done';

export type BestOf = 'BO1' | 'BO3' | 'BO5';

export interface Game {
  id: GameId;
  name: string;
  /** Nom court affiché dans les filtres ("LoL", "CS2"…) */
  short: string;
  /** Tag majuscule ("VAL", "DOTA"…) */
  tag: string;
  /** URL de l'image de fond de la carte jeu */
  bgUrl: string;
  /** Icône/logo du jeu (chemin local, ex. /games/icons/val.svg) */
  logoUrl?: string;
  /** Logo horizontal complet (chemin local, ex. /games/full-logo/val.svg) */
  fullLogoUrl?: string;
}

export interface Team {
  id: string;
  name: string;
  tag: string;
  /** Code pays ISO 3166-1 alpha-2 ("FR", "KR"…) */
  countryCode: string;
  /** URL du logo Liquipedia (Special:FilePath) — absent si pas encore enrichi */
  logoUrl?: string;
}

/** Joueur d'un roster d'équipe (ingéré depuis Liquipedia) */
export interface Player {
  id: string;
  /** Pseudo in-game affiché */
  name: string;
  /** Code pays ISO 3166-1 alpha-2 ("FR", "KR"…) */
  countryCode: string;
  /** Poste/rôle ("Duelist", "Mid", "IGL"…) — absent si inconnu */
  role?: string;
}

/** Lien de diffusion d'un match (Twitch, YouTube…) */
export interface Stream {
  /** Plateforme lisible ("Twitch", "YouTube"…) */
  platform: string;
  url: string;
}

/** Stats d'un joueur sur une carte (scoreboard) */
export interface MapPlayer {
  /** Côté du joueur : 'a' = teamA, 'b' = teamB */
  side: 'a' | 'b';
  name: string;
  countryCode: string;
  kills: number;
  deaths: number;
  assists: number;
  /** Average Combat Score (Valorant) */
  acs?: number;
  /** Average Damage per Round (Valorant / CS2) */
  adr?: number;
  /** % de headshots (Valorant) */
  hs?: number;
  /** Agent joué (Valorant) */
  agent?: string;
  /** Champion joué (League of Legends) */
  champion?: string;
  /** Poste/rôle ("Top", "Mid", "Support"…) — League of Legends */
  role?: string;
  /** Héros joué (Dota 2) */
  hero?: string;
  /** Gold par minute (Dota 2) */
  gpm?: number;
  /** XP par minute (Dota 2) */
  xpm?: number;
  /** Last hits / CS (Dota 2) */
  lasthits?: number;
  /** Net worth / or total (Dota 2) */
  networth?: number;
  /** Niveau du héros (Dota 2) */
  level?: number;
}

/** Une étape du veto des cartes (ban / pick / decider) */
export interface VetoStep {
  order: number;
  type: 'ban' | 'pick' | 'decider';
  /** Équipe qui agit ('a'/'b') ; absent pour la decider */
  team?: 'a' | 'b';
  map: string;
}

/** Score d'une carte (map) jouée ou en cours dans un match */
export interface MapScore {
  name: string;
  scoreA: number;
  scoreB: number;
  winner?: 'a' | 'b';
  live?: boolean;
  /** Scoreboard par joueur (si disponible, ex. Valorant) */
  players?: MapPlayer[];
}

export interface Match {
  id: string;
  gameId: GameId;
  /** L'API embarque les équipes dans la réponse match */
  teamA: Team;
  teamB: Team;
  status: MatchStatus;
  /** Phase du tournoi ("Quart de finale", "Grande Finale"…) */
  phase: string;
  bestOf: BestOf;
  /** Date locale du match, ISO "YYYY-MM-DD" */
  date: string;
  /** Heure locale "HH:mm" */
  time: string;
  scoreA?: number;
  scoreB?: number;
  maps?: MapScore[];
  /** Libellé de la carte en cours, ex. "Lotus · 11–9" (live uniquement) */
  currentMapLabel?: string;
  /** Audience approximative, ex. "52K" (live uniquement) */
  viewers?: string;
  /** Liens de diffusion (Twitch, YouTube…) — absent si non fourni */
  streams?: Stream[];
  /** Veto des cartes (ban/pick/decider) — absent si non fourni */
  veto?: VetoStep[];
  /** Signal best-effort : victoire administrative / forfait probable */
  likelyForfeit?: boolean;
  /** % de la communauté qui pronostique l'équipe A gagnante */
  oddsA?: number;
}
